python/pyspark/sql/pandas/serializers.py

```diff
@@ -194,36 +193,59 @@ def create_array(s, dt: DataType, t: pa.DataType):
                     raise e
             return array
 
-        arrs = []
-        for s, dt in series:
-            t = to_arrow_type(dt) if isinstance(dt, DataType) else dt
-            if t is not None and pa.types.is_struct(t) and not isinstance(dt, UserDefinedType):
-                if not isinstance(s, pd.DataFrame):
-                    raise ValueError("A field of type StructType expects a pandas.DataFrame, "
-                                     "but got: %s" % str(type(s)))
+        def create_arrs_names(s, t: pa.DataType, dt: Optional[DataType] = None):
+            # If input s is empty with zero columns, return empty Arrays with struct
+            if len(s) == 0 and len(s.columns) == 0:
+                return [(pa.array([], type=field.type), field.name) for field in t]
 
-                # Input partition and result pandas.DataFrame empty, make empty Arrays with struct
-                if len(s) == 0 and len(s.columns) == 0:
-                    arrs_names = [(pa.array([], type=field.type), field.name) for field in t]
+            if self._assign_cols_by_name and any(isinstance(name, str) for name in s.columns):
                 # Assign result columns by schema name if user labeled with strings
-                elif self._assign_cols_by_name and any(isinstance(name, str)
-                                                       for name in s.columns):
-                    arrs_names = [
-                        (create_array(s[field.name], struct_field.dataType, field.type), field.name)
+                by_field_name = True
+            else:
+                # Assign result columns by  position
+                by_field_name = False
+
+            if dt is None:
+                if by_field_name:
+                    return [(create_array(s[field.name], field.type), field.name) for field in t]
+                else: 
+                    return [
+                        (create_array(s[s.columns[i]], field.type), field.name)
+                        for i, field in enumerate(t)
+                    ]
+            else:
```

Comment:
`zip(t, dt.fields)` should only be used when t and dt is not the same.

when t = dt, just set dt to None.